src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/collectors/ShardStateCollector.java

```diff
@@ -0,0 +1,132 @@
+package com.amazon.opendistro.elasticsearch.performanceanalyzer.collectors;
+
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.ESResources;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.metrics.AllMetrics;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.metrics.MetricsConfiguration;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.metrics.MetricsProcessor;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.metrics.PerformanceAnalyzerMetrics;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.routing.ShardRouting;
+import org.elasticsearch.cluster.routing.ShardRoutingState;
+import org.jooq.tools.StringUtils;
+
+import static com.amazon.opendistro.elasticsearch.performanceanalyzer.metrics.PerformanceAnalyzerMetrics.SHARD_PRIMARY;
+import static com.amazon.opendistro.elasticsearch.performanceanalyzer.metrics.PerformanceAnalyzerMetrics.SHARD_REPLICA;
+
+public class ShardStateCollector extends PerformanceAnalyzerMetricsCollector implements MetricsProcessor {
+    public static final int SAMPLING_TIME_INTERVAL = MetricsConfiguration.CONFIG_MAP.get(ShardStateCollector.class).samplingInterval;
+    private static final Logger LOG = LogManager.getLogger(ShardStateCollector.class);
+    private static final int KEYS_PATH_LENGTH = 0;
+    private StringBuilder value;
+
+    public ShardStateCollector() {
+        super(SAMPLING_TIME_INTERVAL, "ShardsStateCollector");
+        value = new StringBuilder();
+    }
+
+    @Override
+    void collectMetrics( long startTime) {
+        if (ESResources.INSTANCE.getClusterService() == null) {
+            return;
+        }
+        ClusterState clusterState = ESResources.INSTANCE.getClusterService().state();
+
+        try {
+            value.setLength(0);
+            value.append(PerformanceAnalyzerMetrics.getJsonCurrentMilliSeconds())
+                    .append(PerformanceAnalyzerMetrics.sMetricNewLineDelimitor);
+            for(ShardRouting shard : clusterState.routingTable().allShards()) {
+                String nodeName = StringUtils.EMPTY;
+                if (shard.assignedToNode()) {
+                    nodeName = clusterState.nodes().get(shard.currentNodeId()).getName();
+                }
+                value
+                        .append(new ShardStateMetrics(
+                                shard.getIndexName(),
```

Comment:
Currently the format in which data is emitted is flat, which includes the node name and index name which happen to be user provided strings. Maybe we should add a node level filter as a predicate so that we get information about shards only on this node and not store the node Name (as it is implicit) and make it hierarchical so that no mater the number of shards, we mention index only one time. So something like this:
// 0 is initializing
// 1 is unassigned
// we are skipping reporting the active shards
{
	"unassigned": {
		"index1": [0, 8, ...],
		"index10": [0]
	},
	"initializing": {
		"index99": [0, 8, ...],
		"index1": [7]
	}
}

Reply:
Changed it to indexName at the parent level. So Sample metric would look like
Sample event :
^shard_state_metrics
{"current_time":1600677426860}
{IndexName:"pmc"}
{"ShardID":2,"ShardType":"primary","NodeName":"elasticsearch2","Shard_State":"Unassigned"}
{"ShardID":2,"ShardType":"primary","NodeName":"elasticsearch2","Shard_State:"Initializing"}
{IndexName:"pmc1"}
{"ShardID":2,"ShardType":"primary","NodeName":"elasticsearch2","Shard_State":"Unassigned"}